File: 02. Estructurales/04. Decorator/01. decorator.js
/***

    Imaginemos que estamos construyendo una aplicación de creación de perfiles de usuario. 
    Tenemos una clase base UserProfile que representa un perfil de usuario básico. 
    Queremos poder agregar funcionalidades adicionales a los perfiles de usuario, 
    como la capacidad de agregar una foto de perfil o la capacidad de agregar una descripción personalizada.

    Aquí es donde el patrón Decorator puede ser útil. 
    Podemos utilizar el patrón Decorator para agregar funcionalidades adicionales 
    a los perfiles de usuario sin modificar la clase base.

*/

// Comencemos definiendo nuestra clase base UserProfile:

class UserProfile {
    constructor(username) {
        this.username = username;
    }

    getProfile() {
        return `Perfil de usuario: ${this.username}`;
    }
}

  // A continuación, definimos nuestros decoradores que agregarán funcionalidades adicionales a los perfiles de usuario:

// Decorador para agregar una foto de perfil
class ProfilePhotoDecorator {
    constructor(userProfile, photoUrl) {
        this.userProfile = userProfile;
        this.photoUrl = photoUrl;
    }

    getProfile() {
        return `${this.userProfile.getProfile()}, Foto de perfil: ${this.photoUrl}`;
    }
}

// Decorador para agregar una descripción personalizada
class CustomDescriptionDecorator {
    constructor(userProfile, description) {
        this.userProfile = userProfile;
        this.description = description;
    }

    getProfile() {
        return `${this.userProfile.getProfile()}, Descripción: ${this.description}`;
    }
}

  // Finalmente, podemos utilizar el patrón Decorator para agregar funcionalidades adicionales a los perfiles de usuario:

// Crear un perfil de usuario básico
const basicProfile = new UserProfile("JohnDoe");
console.log(basicProfile.getProfile()); // Perfil de usuario: JohnDoe

// Agregar una foto de perfil al perfil básico
const profileWithPhoto = new ProfilePhotoDecorator(basicProfile, "https://example.com/photo.jpg");
console.log(profileWithPhoto.getProfile()); // Perfil de usuario: JohnDoe, Foto de perfil: https://example.com/photo.jpg

// Agregar una descripción personalizada al perfil básico
const profileWithDescription = new CustomDescriptionDecorator(basicProfile, "¡Hola! Soy John Doe.");
console.log(profileWithDescription.getProfile()); // Perfil de usuario: JohnDoe, Descripción: ¡Hola! Soy John Doe.

/***

    En este ejemplo, el patrón Decorator nos permite agregar funcionalidades adicionales a los perfiles de usuario 
    sin modificar la clase base UserProfile. Los decoradores ProfilePhotoDecorator y CustomDescriptionDecorator 
    envuelven la clase base y agregan la funcionalidad adicional deseada.

*/

File: 02. Estructurales/04. Decorator/01. decorator.py
# Definimos nuestra clase base UserProfile
class UserProfile:
    def __init__(self, username):
        self.username = username

    def get_profile(self):
        return f"Perfil de usuario: {self.username}"

# Definimos nuestro decorador para agregar una foto de perfil
class ProfilePhotoDecorator:
    def __init__(self, user_profile, photo_url):
        self.user_profile = user_profile
        self.photo_url = photo_url

    def get_profile(self):
        return f"{self.user_profile.get_profile()}, Foto de perfil: {self.photo_url}"

# Definimos nuestro decorador para agregar una descripción personalizada
class CustomDescriptionDecorator:
    def __init__(self, user_profile, description):
        self.user_profile = user_profile
        self.description = description

    def get_profile(self):
        return f"{self.user_profile.get_profile()}, Descripción: {self.description}"

# Utilizamos el patrón Decorator para agregar funcionalidades adicionales a los perfiles de usuario

# Creamos un perfil de usuario básico
basic_profile = UserProfile("JohnDoe")
print(basic_profile.get_profile()) # Perfil de usuario: JohnDoe

# Agregamos una foto de perfil al perfil básico
profile_with_photo = ProfilePhotoDecorator(basic_profile, "https://example.com/photo.jpg")
print(profile_with_photo.get_profile()) # Perfil de usuario: JohnDoe, Foto de perfil: https://example.com/photo.jpg

# Agregamos una descripción personalizada al perfil básico
profile_with_description = CustomDescriptionDecorator(basic_profile, "¡Hola! Soy John Doe.")
print(profile_with_description.get_profile()) # Perfil de usuario: JohnDoe, Descripción: ¡Hola! Soy John Doe.